src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	UINT32				tile_height;			/* height of video tiles */
+	UINT32				x_tiles;				/* number of tiles in the X direction */
+
+	UINT8				vblank;					/* VBLANK state */
+	bool				vblank_dont_swap;		/* don't actually swap when we hit this point */
+	bool				vblank_flush_pending;
+
+	/* triangle setup info */
+	INT16				ax, ay;					/* vertex A x,y (12.4) */
+	INT16				bx, by;					/* vertex B x,y (12.4) */
+	INT16				cx, cy;					/* vertex C x,y (12.4) */
+	INT32				startr, startg, startb, starta; /* starting R,G,B,A (12.12) */
+	INT32				startz;					/* starting Z (20.12) */
+	INT64				startw;					/* starting W (16.32) */
+	INT32				drdx, dgdx, dbdx, dadx;	/* delta R,G,B,A per X */
+	INT32				dzdx;					/* delta Z per X */
+	INT64				dwdx;					/* delta W per X */
+	INT32				drdy, dgdy, dbdy, dady;	/* delta R,G,B,A per Y */
+	INT32				dzdy;					/* delta Z per Y */
+	INT64				dwdy;					/* delta W per Y */
+
+	stats_block			lfb_stats;				/* LFB-access statistics */
+
+	UINT8				sverts;					/* number of vertices ready */
+	setup_vertex		svert[3];				/* 3 setup vertices */
+
+	fifo_state			fifo;					/* framebuffer memory fifo */
+
+	UINT8				fogblend[64];			/* 64-entry fog table */
+	UINT8				fogdelta[64];			/* 64-entry fog table */
+	UINT8				fogdelta_mask;			/* mask for for delta (0xff for V1, 0xfc for V2) */
+
+	//rgb_t				clut[512];				/* clut gamma data */
+};
+
+struct dac_state
+{
+	UINT8				reg[8];					/* 8 registers */
+	UINT8				read_result;			/* pending read result */
+};
+
+#ifdef C_ENABLE_VOODOO_OPENGL
+#ifndef GLhandleARB
+#ifdef __APPLE__
+typedef void* GLhandleARB;
+#else
+typedef unsigned int GLhandleARB;
+#endif /* __APPLE__ */
+#endif /* GLhandleARB */
+
+struct raster_info
+{
+	raster_info			*next;					/* pointer to next entry with the same hash */
+#ifdef C_ENABLE_VOODOO_DEBUG
+	bool				is_generic;				/* true if this is one of the generic rasterizers */
+	UINT8				display;				/* display index */
+	UINT32				hits;					/* how many hits (pixels) we've used this for */
+	UINT32				polys;					/* how many polys we've used this for */
+#endif
+	UINT32				eff_color_path;			/* effective fbzColorPath value */
+	UINT32				eff_alpha_mode;			/* effective alphaMode value */
+	UINT32				eff_fog_mode;			/* effective fogMode value */
+	UINT32				eff_fbz_mode;			/* effective fbzMode value */
+	UINT32				eff_tex_mode_0;			/* effective textureMode value for TMU #0 */
+	UINT32				eff_tex_mode_1;			/* effective textureMode value for TMU #1 */
+
+	bool				shader_ready;
+	GLhandleARB			so_shader_program;
+	GLhandleARB			so_vertex_shader;
+	GLhandleARB			so_fragment_shader;
+	INT32*				shader_ulocations;
+};
+#endif
+
+struct draw_state
+{
+	double frame_start;
+	float vfreq;
+	bool override_on;
+	bool screen_update_requested;
+	bool screen_update_pending;
+};
+
+struct triangle_worker
+{
+	bool threads_active, use_threads, disable_bilinear_filter;
+	UINT16 *drawbuf;
+	poly_vertex v1, v2, v3;
+	INT32 v1y, v3y, totalpix;
+	SDL_sem* sembegin[TRIANGLE_THREADS];
```

Comment:
Ooops, what do we need semaphors for? Multithreading? Ideally, I'd like this to be framework agnostic (so no SDL references).

Reply:
Yup, there are 4 worker threads IIRC.

Reply:
I think initially this code was implemented using POSIX threads, then DOSBox team asked to rewrite it using SDL_Thread. I think we'll end up switching to `std::thread` here... But not touching this now - switch should be its own separate PR.